### PR TITLE
Refactor the certificate and kubeconfig code in the kubeadm binary into two phases

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

First stab at refactoring kubeadm code into logically independent phases.
This defines two phases in the kubeadm init process:
 - certs: Takes some API values as input (the API will be refactored in a later PR), and generates certificates in the pki directory
 - kubeconfig: Takes the pki directory and the endpoint where the master is located and produces two kubeconfig files: admin.conf and kubelet.conf

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:
Required long-term for graduating our API

**Special notes for your reviewer**:

### Old sample output
The earlier kubeconfig code had a bug in it; see this example:
_admin.conf:_
```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <data>
    server: https://192.168.200.x:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: admin
  name: admin@kubernetes
- context:
    cluster: kubernetes
    user: kubelet
  name: kubelet@kubernetes
current-context: admin@kubernetes
kind: Config
preferences: {}
users:
- name: admin
  user:
    client-certificate-data: <data>
    client-key-data: <data>
- name: kubelet
  user:
    client-certificate-data: <data>
    client-key-data: <data>
```
kubelet.conf:
```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <data>
    server: https://192.168.200.x:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: admin
  name: admin@kubernetes
- context:
    cluster: kubernetes
    user: kubelet
  name: kubelet@kubernetes
current-context: admin@kubernetes
kind: Config
preferences: {}
users:
- name: admin
  user:
    client-certificate-data: <data>
    client-key-data: <data>
- name: kubelet
  user:
    client-certificate-data: <data>
    client-key-data: <data>
```
```console
$ shasum /etc/kubernetes/*.conf
2b22b25cc4c97e5619ece6c43badf42b87c4970a  /etc/kubernetes/admin.conf
2b22b25cc4c97e5619ece6c43badf42b87c4970a  /etc/kubernetes/kubelet.conf
```

#### New output
admin.conf
```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <data>
    server: https://192.168.200.x:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: admin
  name: admin@kubernetes
current-context: admin@kubernetes
kind: Config
preferences: {}
users:
- name: admin
  user:
    client-certificate-data: <data>
    client-key-data: <data>
```
kubelet.conf
```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <data>
    server: https://192.168.200.x:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: kubelet
  name: kubelet@kubernetes
current-context: kubelet@kubernetes
kind: Config
preferences: {}
users:
- name: kubelet
  user:
    client-certificate-data: <data>
    client-key-data: <data>
```

**Release note**:

```release-note
Refactor the certificate and kubeconfig code in the kubeadm binary into two phases
```

PTAL @dgoodwin @jbeda @mikedanese @errordeveloper @pipejakob @lukemarsden

### CODE REVIEW

File: pkg/util/cert/io.go
Comment (mikedanese) [inline pkg/util/cert/io.go]: This is a weird check. It doesn't seem necessary.

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: Do the err check separately so we don't print couldnt read file because: <nil>

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: We support ecdsa everywhere but here?

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: just `return WriteKubeconfigToDisk(filepath, config)`

Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: fixed

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: added a TODO for this and a comment why it can't be supported right away
deferring to the future

File: pkg/util/cert/io.go
Comment (luxas) [inline pkg/util/cert/io.go]: fixed, I didn't understand why this was present in CertsFromFile from where I took it, removed now

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go
Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: will do

Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: at a closer look, err will be non-nil if `len(caCerts) == 0` so that check didn't make sense

**Comment by luxas** [comment]
@mikedanese Updated and comments fixed.
PTAL

Would be great if you could approve this so I can continue working on this refactoring job tomorrow and next week.

**Comment by luxas** [comment]
Hmm, @mikedanese is the bazel build failing constantly now?

**Comment by luxas** [comment]
Everything but bazel passed, please retest when you've got it working again

File: cmd/kubeadm/app/cmd/init.go
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Why was this removed? Doesn't seem related to the PR title and description.

File: cmd/kubeadm/app/cmd/init.go
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Can we move filename to a constant?

File: cmd/kubeadm/app/cmd/join.go
Comment (pires) [inline cmd/kubeadm/app/cmd/join.go]: Can we move filename to a constant?

**Review by pires — CHANGES_REQUESTED** [review]
Overall LGTM.

**Comment by mikedanese** [comment]
@k8s-bot bazel test this

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: This should have been done in an earlier PR, but it seems like it was missed.
This will and should only be doable from the API

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Will do in a follow up

File: cmd/kubeadm/app/cmd/join.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: Will do in a follow up

**Comment by mikedanese** [comment]
What do you envision happening in cmd/kubeadm/phases/? Can you name some other functionalities that will move into a phases/ subfolder?

**Comment by pires** [comment]
> What do you envision happening in cmd/kubeadm/phases/? Can you name some other functionalities that will move into a phases/ subfolder?

Are phases supposed to eventually be extracted to commands other binaries can re-use?

**Comment by luxas** [comment]
@mikedanese I talked about this in the call today.
Basically these phases (copied from the meeting notes):

Identification of the kubeadm phases so that we can define an "API" (interface) (configfile + commandline) so that we can split kubeadm into separate independently-runnable phases, rough outline:
 - 1: Generating certificates
 - 2: Generating the admin.conf and the temporary kubelet.conf for bootstrapping the kubelet
 - 3: Bootstrapping the control plane
 - 4: API Configuration
 - 5: Discovery
 - 6: Create essential addons

This makes kubeadm ~support running the first two individually (but no CLI support added yet or API changes, that will come later)

Also see https://github.com/kubernetes/kubeadm/pull/100 once I have had time to write things there.

However, this is only the first PR in a series of a few for this to happen. This is just work for making it possible to run individual phases in the future, code changes that have to happen and cleanup generally in the codebase.

I would be grateful if you lgtm'd this now so I can move forward, right now I'm blocked on this.

**Comment by luxas** [comment]
@pires Yes, this is the initial code that will make that possible.
Now the code in the codebase doesn't talk with each other back and forth, instead the code that does the individual phase is isolated in a package from the rest. It also makes the source more understandable long-term

File: cmd/kubeadm/app/cmd/init.go
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/init.go]: Why not now? It seems like a small enough fix.

**Comment by mikedanese** [comment]
This looks fine but I think you should just address @pires comments since they are so small.

**Comment by pires** [comment]
@luxas great, you already have my LGTM.

**Comment by mikedanese** [comment]
can you rebase and run gofmt and update-bazel

**Comment by luxas** [comment]
Rebased, gofmt'd and bazel updated

**Pull request merged**
